profile/profile_test.go

```diff
@@ -912,6 +945,134 @@ func locationHash(s *Sample) string {
 	return tb
 }
 
+func TestHasTag(t *testing.T) {
+	var testcases = []struct {
+		desc       string
+		sample     *Sample
+		key        string
+		value      string
+		wantHasTag bool
+	}{
+		{
+			desc: "empty label does not have tag",
+			sample: &Sample{
+				Location: []*Location{cpuL[0]},
+				Value:    []int64{1000, 1000},
+			},
+			key:        "key",
+			value:      "value",
+			wantHasTag: false,
+		},
+		{
+			desc: "label with one key and value has tag",
+			sample: &Sample{
+				Location: []*Location{cpuL[0]},
+				Value:    []int64{1000, 1000},
+				Label:    map[string][]string{"key": {"value"}},
+			},
+			key:        "key",
+			value:      "value",
+			wantHasTag: true,
+		},
+		{
+			desc: "label with one key and value does not have tag",
+			sample: &Sample{
+				Location: []*Location{cpuL[0]},
+				Value:    []int64{1000, 1000},
+				Label:    map[string][]string{"key": {"value"}},
+			},
+			key:        "key1",
+			value:      "value1",
+			wantHasTag: false,
+		},
+		{
+			desc: "label with many keys and values has tag",
+			sample: &Sample{
+				Location: []*Location{cpuL[0]},
+				Value:    []int64{1000, 1000},
+				Label: map[string][]string{
+					"key1": {"value2", "value1"},
+					"key2": {"value1", "value2", "value2"},
+					"key3": {"value1", "value2", "value2"},
+				},
+			},
+			key:        "key1",
+			value:      "value1",
+			wantHasTag: true,
+		},
+		{
+			desc: "label with many keys and values does not have tag",
+			sample: &Sample{
+				Location: []*Location{cpuL[0]},
+				Value:    []int64{1000, 1000},
+				Label: map[string][]string{
+					"key1": {"value2", "value1"},
+					"key2": {"value1", "value2", "value2"},
+					"key3": {"value1", "value2", "value2"},
+				},
+			},
+			key:        "key5",
+			value:      "value5",
+			wantHasTag: false,
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.desc, func(t *testing.T) {
+			if gotHasTag := tc.sample.HasTag(tc.key, tc.value); gotHasTag != tc.wantHasTag {
+				t.Errorf("sample.HasTag(%q, %q) got %v, want %v", tc.key, tc.value, gotHasTag, tc.wantHasTag)
+			}
+		})
+	}
+}
+
+func TestSetTag(t *testing.T) {
+	var testcases = []struct {
+		desc     string
+		profile  *Profile
+		setKey   string
+		setVal   []string
+		wantTags []map[string][]string
+	}{
+		{
+			desc:    "some samples have tag already",
```

Comment:
Did you mean to add more test cases? E.g. what if sample doesn't have the tags.

Reply:
Done.
Also did some refactoring of this test.